### PR TITLE
Fixed bug in entropy calculation and ZPE calculation

### PR DESCRIPTION
**Pull request by KristofDeWispelaere**
Yay! It works :D

### CODE REVIEW

File: __init__.py [hunk 1]
```diff
@@ -376,21 +387,29 @@ def get_harmonic_thermo(self, hnu):
...
+        entropic_energy_hmodes = [] #entropy contribution per mode
+        x = 0.
 
         for e in hnu:
             if e.imag == 0 and e.real >= 0.010:
                 Z_mode = 1./(1.-np.exp(-e.real/(self.kT)))
                 ZPE = 0.5 * e.real
                 e_min_exitation = e.real
+                x = e_min_exitation/self.kT
+                entropic_energy_hmodes.append(self.kT*(x/(np.exp(x)-1.) - np.log(1.-np.exp(-x))))
```
Comment (keldLundgaard) [inline __init__.py]: Lines should be maximum 79 characters long. 
See https://www.python.org/dev/peps/pep-0008/#maximum-line-length

File: __init__.py [hunk 2]
```diff
@@ -329,16 +329,22 @@ def calculate_anharmonic_thermo(self):
...
         self.Z_modes = []  # partition function for mode
         self.e_exitation_modes = []  # principle energy of mode
+        self.entropic_energy_modes = [] #entropy contribution per vibrational mode
+        x = 0.
+        y = 0.
 
         for i, an_mode in enumerate(self.an_modes):
             # Partition function of mode
             Z_mode = 0.
+            y = 0.
```
Comment (keldLundgaard) [inline __init__.py]: What is y? Please use a more descriptive?

File: __init__.py [hunk 3]
```diff
@@ -350,17 +356,22 @@ def calculate_anharmonic_thermo(self):
...
+                x = e_min_exitation/self.kT
+                self.entropic_energy_modes.append(self.kT*(x/(np.exp(x)-1.) - np.log(1.-np.exp(-x))))
+
             else:
                 ZPE = 0.
                 Z_mode = 1.
                 e_min_exitation = 0.
+                self.entropic_energy_modes.append(0.)
 
+            x = 0.
```
Comment (keldLundgaard) [inline __init__.py]: Also a better name for x

File: anh_base.py [hunk 4]
```diff
@@ -136,7 +138,9 @@ def get_thermo(self, fitobj):
 
             Hcoeff = units._hbar**2/(2*units._amu * units._e * 1e-20)
 
+            #groundstate_energy = self.an_mode['displacement_energies'][0]
```
Comment (keldLundgaard) [inline anh_base.py]: Please delete the commented line

File: anh_base.py [hunk 5]
```diff
@@ -119,14 +119,16 @@ def get_thermo(self, fitobj):
             xmin = 0.
             xmax = xmin+2.*np.pi/self.an_mode['symnumber']
 
-            groundstate_energy = self.an_mode['displacement_energies'][0]
+            #groundstate_energy = self.an_mode['displacement_energies'][0]
```
Comment (keldLundgaard) [inline anh_base.py]: Delete commented line

File: __init__.py [hunk 6]
```diff
@@ -376,21 +387,29 @@ def get_harmonic_thermo(self, hnu):
...
                 Z_mode = 1.
                 e_min_exitation = 0.
+                entropic_energy_hmodes.append(0.)
             ZPE_hmodes.append(ZPE)
             Z_hmodes.append(Z_mode)
             e_excitation_hmodes.append(e_min_exitation)
+            x = 0.
 
-        entropic_energy_hmodes = -1*self.kT*np.log(Z_hmodes)
+        #entropic_energy_hmodes = -1*self.kT*np.log(Z_hmodes)
```
Comment (keldLundgaard) [inline __init__.py]: Delete commented line

File: __init__.py [hunk 7]
```diff
@@ -350,17 +356,22 @@ def calculate_anharmonic_thermo(self):
...
+                self.entropic_energy_modes.append(0.)
 
+            x = 0.
             self.ZPE_modes.append(ZPE)
             self.Z_modes.append(Z_mode)
             self.e_exitation_modes.append(e_min_exitation)
 
         # Calculating the entropic energy for each mode
-        self.entropic_energy_modes = -1*self.kT*np.log(self.Z_modes)
+        #self.entropic_energy_modes = -1*self.kT*np.log(self.Z_modes)
```
Comment (keldLundgaard) [inline __init__.py]: Delete commented line

File: __init__.py [hunk 8]
```diff
@@ -329,16 +329,22 @@ def calculate_anharmonic_thermo(self):
...
             for ei in an_mode['energy_levels']:
                 Z_mode += np.exp(-(ei-an_mode['ZPE'])/self.kT)
+                y += np.exp(-(ei-an_mode['ZPE'])/self.kT) * (ei-an_mode['ZPE'])
+
+            self.entropic_energy_modes.append(self.kT * np.log(Z_mode) + 1./Z_mode * y)
 
             # Difference between ZPE and first excited energy level
-            e_min_exitation = (
-                an_mode['energy_levels'][1]-an_mode['energy_levels'][0])
+            e_min_exitation = (an_mode['energy_levels'][1]-an_mode['energy_levels'][0])
```
Comment (keldLundgaard) [inline __init__.py]: Probably needs to be 2 lines with 79 max line length

File: __init__.py [hunk 9]
```diff
@@ -1,5 +1,5 @@
 """Anharmonic modes module for ASE
-
+#KDW
```
Comment (keldLundgaard) [inline __init__.py]: Delete

File: __init__.py (hunk 3 above)
Comment (KristofDeWispelaere) [inline __init__.py]: I copied this from the HarmonicThermo module of ASE. I know x doesn't mean anything; any suggestion how to call E/kT?

File: __init__.py [hunk 10]
```diff
@@ -329,16 +329,22 @@ def calculate_anharmonic_thermo(self):
         self.ZPE_modes = []  # Zero point energy for mode
         self.Z_modes = []  # partition function for mode
         self.e_exitation_modes = []  # principle energy of mode
+        self.entropic_energy_modes = [] #entropy contribution per vibrational mode
+        x = 0.
+        y = 0.
```
Comment (keldLundgaard) [inline __init__.py]: Seems like x and y can be deleted here. Please double check.

File: __init__.py (hunk 3 above)
Comment (keldLundgaard) [inline __init__.py]: x = 0. can also be deleted here

File: __init__.py [hunk 11]
```diff
@@ -376,21 +387,29 @@ def get_harmonic_thermo(self, hnu):
         ZPE_hmodes = []  # Zero point energy for mode
         Z_hmodes = []  # partition function for mode
         e_excitation_hmodes = []  # principle energy of mode
+        entropic_energy_hmodes = [] #entropy contribution per mode
+        x = 0.
```
Comment (keldLundgaard) [inline __init__.py]: Delete x = 0.

File: __init__.py [hunk 12]
```diff
@@ -376,21 +387,29 @@ def get_harmonic_thermo(self, hnu):
...
+
             else:
                 ZPE = 0.
                 Z_mode = 1.
                 e_min_exitation = 0.
+                entropic_energy_hmodes.append(0.)
             ZPE_hmodes.append(ZPE)
             Z_hmodes.append(Z_mode)
             e_excitation_hmodes.append(e_min_exitation)
+            x = 0.
```
Comment (keldLundgaard) [inline __init__.py]: delete

File: anh_base.py [hunk 13]
```diff
@@ -119,14 +119,16 @@ def get_thermo(self, fitobj):
...
+            groundstate_energy = min(self.an_mode['displacement_energies'])
 
         elif self.an_mode['type'] == 'vibration':
             Hcoeff = units._hbar**2 / (2.*units._amu * units._e * 1e-20)
 
             xmin = np.min(self.an_mode['displacements'])
             xmax = np.max(self.an_mode['displacements'])
 
+            #groundstate_energy = self.an_mode['displacement_energies'][0]
```
Comment (keldLundgaard) [inline anh_base.py]: remove comment

File: __init__.py [hunk 14]
```diff
@@ -329,16 +329,23 @@ def calculate_anharmonic_thermo(self):
...
             for ei in an_mode['energy_levels']:
                 Z_mode += np.exp(-(ei-an_mode['ZPE'])/self.kT)
+                dZ_dT += np.exp(-(ei-an_mode['ZPE'])/self.kT) * (ei-an_mode['ZPE'])
+
+            self.entropic_energy_modes.append(self.kT * np.log(Z_mode) + 1./Z_mode * dZ_dT)
 
             # Difference between ZPE and first excited energy level
             e_min_exitation = (
-                an_mode['energy_levels'][1]-an_mode['energy_levels'][0])
+                    an_mode['energy_levels'][1]-an_mode['energy_levels'][0])
```
Comment (keldLundgaard) [inline __init__.py]: Please remove extra indenting

File: __init__.py [hunk 15]
```diff
@@ -350,18 +357,21 @@ def calculate_anharmonic_thermo(self):
...
+                self.entropic_energy_modes.append(
+                        self.kT*(E_kT/(np.exp(E_kT)-1.) - np.log(1.-np.exp(-E_kT))))
+
             else:
                 ZPE = 0.
                 Z_mode = 1.
                 e_min_exitation = 0.
+                self.entropic_energy_modes.append(0.)
 
+            E_kT = 0.
```
Comment (keldLundgaard) [inline __init__.py]: Delete here! Right?

File: __init__.py [hunk 16]
```diff
@@ -376,21 +386,28 @@ def get_harmonic_thermo(self, hnu):
         ZPE_hmodes = []  # Zero point energy for mode
         Z_hmodes = []  # partition function for mode
         e_excitation_hmodes = []  # principle energy of mode
+        entropic_energy_hmodes = [] #entropy contribution per mode
+        E_kT = 0.
```
Comment (keldLundgaard) [inline __init__.py]: Delete here! Right?

File: __init__.py [hunk 17]
```diff
@@ -376,21 +386,28 @@ def get_harmonic_thermo(self, hnu):
...
+
             else:
                 ZPE = 0.
                 Z_mode = 1.
                 e_min_exitation = 0.
+                entropic_energy_hmodes.append(0.)
             ZPE_hmodes.append(ZPE)
             Z_hmodes.append(Z_mode)
             e_excitation_hmodes.append(e_min_exitation)
+            E_kT = 0.
```
Comment (keldLundgaard) [inline __init__.py]: Delete here! Right?

File: __init__.py [hunk 18]
```diff
@@ -350,18 +357,21 @@ def calculate_anharmonic_thermo(self):
                 Z_mode = 1./(1.-np.exp(-e.real/(self.kT)))
                 ZPE = 0.5 * e.real
                 e_min_exitation = e.real
+                E_kT = e_min_exitation/self.kT
```
Comment (keldLundgaard) [inline __init__.py]: Now E_kt (x) is a very local thing only defined here and used on the next line so it is fine to call it x.

File: __init__.py [hunk 19]
```diff
@@ -329,16 +329,23 @@ def calculate_anharmonic_thermo(self):
         self.ZPE_modes = []  # Zero point energy for mode
         self.Z_modes = []  # partition function for mode
         self.e_exitation_modes = []  # principle energy of mode
+        self.entropic_energy_modes = [] #entropy contribution per vibrational mode
+        E_kT = 0.
```
Comment (keldLundgaard) [inline __init__.py]: Delete here! Right?

File: __init__.py [hunk 20]
```diff
@@ -329,16 +329,23 @@ def calculate_anharmonic_thermo(self):
...
+        dZ_dT = 0.
```
Comment (keldLundgaard) [inline __init__.py]: Delete here! Right?

File: __init__.py [hunk 21]
```diff
@@ -329,16 +329,23 @@ def calculate_anharmonic_thermo(self):
...
 
         for i, an_mode in enumerate(self.an_modes):
             # Partition function of mode
             Z_mode = 0.
+            dZ_dT = 0.
```
Comment (keldLundgaard) [inline __init__.py]: I like that name :)

File: __init__.py [hunk 22]
```diff
@@ -376,21 +386,28 @@ def get_harmonic_thermo(self, hnu):
...
         e_excitation_hmodes = []  # principle energy of mode
+        entropic_energy_hmodes = [] #entropy contribution per mode
+        E_kT = 0.
 
         for e in hnu:
             if e.imag == 0 and e.real >= 0.010:
                 Z_mode = 1./(1.-np.exp(-e.real/(self.kT)))
                 ZPE = 0.5 * e.real
                 e_min_exitation = e.real
+                E_kT = e_min_exitation/self.kT
```
Comment (keldLundgaard) [inline __init__.py]: Same as earlier. x is better than E_kT as I think calling it E_kT makes it more confusing than x.

File: __init__.py [hunk 23]
```diff
@@ -376,21 +386,28 @@ def get_harmonic_thermo(self, hnu):
...
+                entropic_energy_hmodes.append(
+                        self.kT*(E_kT/(np.exp(E_kT)-1.) - np.log(1.-np.exp(-E_kT))))
```
Comment (keldLundgaard) [inline __init__.py]: Need spacing around -, *, + 

4+3  should be 4 + 3

**Review by keldLundgaard — APPROVED** [review]

**Pull request merged**